### ISSUE TITLE
[PSyIR] Unused variables in adjointed code

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#1962]
Occasionally the adjointed kernel code produced by PSyAD contains (loop?) variables that are declared but unused. Although this is fine, the LFRic build environment flags such variables as errors. It would therefore be better if we didn't create such variables. This could be done by the backend although care would have to be taken if any CodeBlock nodes are present. Does @sergisiso agree?

**Comment by sergisiso** [comment]
It can be done but it is a bit risky. Maybe could be a optional backend parameter for now?
Beside recursing the References and CodeBlocks you will also need to check in the other declarations UnknownTypes and constant_values.

**Comment by arporter** [comment]
Yes, it does feel a bit risky. Even for my particular use case (where the variables that are causing trouble are simple and not in any codeblocks), there could be other variables that we don't fully handle. The more I think about it, the more I'm inclined to think we should work out which transformation it is that's causing a variable to become un-used and fix it there.
